From 5536d2c238a094dfe2b94f56b8e9d90168ec7c4d Mon Sep 17 00:00:00 2001
From: Dinesh Karthikeyan <quic_dinek@quicinc.com>
Date: Thu, 3 Nov 2022 15:20:19 +0530
Subject: [PATCH] ath12k: Add MLO caps

Add support to fill mlo parameters in the host capability send message and to
send the mlo capability to FW.

Signed-off-by: Dinesh Karthikeyan <quic_dinek@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h |   1 +
 drivers/net/wireless/ath/ath12k/qmi.c  | 179 +++++++++++++++++++++++++
 drivers/net/wireless/ath/ath12k/qmi.h  |  20 ++-
 3 files changed, 199 insertions(+), 1 deletion(-)

Index: backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/core.h
===================================================================
--- backports-20220404-5.4.164-f40abb4788.orig/drivers/net/wireless/ath/ath12k/core.h	2022-12-21 11:34:10.775816102 +0530
+++ backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/core.h	2022-12-21 11:34:10.767816089 +0530
@@ -189,6 +189,7 @@ struct ath12k_he {
 #define ATH12K_MAX_SOCS		3
 #define ATH12K_GROUP_MAX_RADIO	(ATH12K_MAX_SOCS * MAX_RADIOS)
 #define ATH12K_INVALID_GRP_ID	0xFF
+#define ATH12K_MAX_MLO_PEER	0x100
 
 enum {
 	WMI_HOST_TP_SCALE_MAX   = 0,
Index: backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/qmi.c
===================================================================
--- backports-20220404-5.4.164-f40abb4788.orig/drivers/net/wireless/ath/ath12k/qmi.c	2022-12-21 11:34:10.775816102 +0530
+++ backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/qmi.c	2022-12-21 11:36:38.540059133 +0530
@@ -210,6 +210,50 @@ static struct qmi_elem_info qmi_wlanfw_q
 	},
 };
 
+static struct qmi_elem_info qmi_wlanfw_host_mlo_chip_info_s_v01_ei[] = {
+	{
+		.data_type      = QMI_UNSIGNED_1_BYTE,
+		.elem_len       = 1,
+		.elem_size      = sizeof(u8),
+		.array_type     = NO_ARRAY,
+		.tlv_type       = 0,
+		.offset         = offsetof(struct qmi_wlanfw_host_mlo_chip_info_s_v01,
+					   chip_id),
+	},
+	{
+		.data_type      = QMI_UNSIGNED_1_BYTE,
+		.elem_len       = 1,
+		.elem_size      = sizeof(u8),
+		.array_type     = NO_ARRAY,
+		.tlv_type       = 0,
+		.offset         = offsetof(struct qmi_wlanfw_host_mlo_chip_info_s_v01,
+					   num_local_links),
+	},
+	{
+		.data_type      = QMI_UNSIGNED_1_BYTE,
+		.elem_len       = QMI_WLANFW_MAX_NUM_MLO_LINKS_PER_CHIP_V01,
+		.elem_size      = sizeof(u8),
+		.array_type     = STATIC_ARRAY,
+		.tlv_type       = 0,
+		.offset         = offsetof(struct qmi_wlanfw_host_mlo_chip_info_s_v01,
+					   hw_link_id),
+	},
+	{
+		.data_type      = QMI_UNSIGNED_1_BYTE,
+		.elem_len       = QMI_WLANFW_MAX_NUM_MLO_LINKS_PER_CHIP_V01,
+		.elem_size      = sizeof(u8),
+		.array_type     = STATIC_ARRAY,
+		.tlv_type       = 0,
+		.offset         = offsetof(struct qmi_wlanfw_host_mlo_chip_info_s_v01,
+					   valid_mlo_link_id),
+	},
+	{
+		.data_type      = QMI_EOTI,
+		.array_type     = NO_ARRAY,
+		.tlv_type       = QMI_COMMON_TLV_TYPE,
+	},
+};
+
 static struct qmi_elem_info qmi_wlanfw_host_cap_req_msg_v01_ei[] = {
 	{
 		.data_type	= QMI_OPT_FLAG,
@@ -536,7 +580,7 @@ static struct qmi_elem_info qmi_wlanfw_h
 					   mlo_capable_valid),
 	},
 	{
-		.data_type	= QMI_OPT_FLAG,
+		.data_type	= QMI_UNSIGNED_1_BYTE,
 		.elem_len	= 1,
 		.elem_size	= sizeof(u8),
 		.array_type	= NO_ARRAY,
@@ -563,6 +607,79 @@ static struct qmi_elem_info qmi_wlanfw_h
 					   mlo_chip_id),
 	},
 	{
+		.data_type      = QMI_OPT_FLAG,
+		.elem_len       = 1,
+		.elem_size      = sizeof(u8),
+		.array_type     = NO_ARRAY,
+		.tlv_type       = 0x23,
+		.offset         = offsetof(struct qmi_wlanfw_host_cap_req_msg_v01,
+					   mlo_group_id_valid),
+	},
+	{
+		.data_type      = QMI_UNSIGNED_1_BYTE,
+		.elem_len       = 1,
+		.elem_size      = sizeof(u8),
+		.array_type     = NO_ARRAY,
+		.tlv_type       = 0x23,
+		.offset         = offsetof(struct qmi_wlanfw_host_cap_req_msg_v01,
+					   mlo_group_id),
+	},
+	{
+		.data_type      = QMI_OPT_FLAG,
+		.elem_len       = 1,
+		.elem_size      = sizeof(u8),
+		.array_type     = NO_ARRAY,
+		.tlv_type       = 0x24,
+		.offset         = offsetof(struct qmi_wlanfw_host_cap_req_msg_v01,
+					   max_mlo_peer_valid),
+	},
+	{
+		.data_type      = QMI_UNSIGNED_2_BYTE,
+		.elem_len       = 1,
+		.elem_size      = sizeof(u16),
+		.array_type     = NO_ARRAY,
+		.tlv_type       = 0x24,
+		.offset         = offsetof(struct qmi_wlanfw_host_cap_req_msg_v01,
+					   max_mlo_peer),
+	},
+	{
+		.data_type      = QMI_OPT_FLAG,
+		.elem_len       = 1,
+		.elem_size      = sizeof(u8),
+		.array_type     = NO_ARRAY,
+		.tlv_type       = 0x25,
+		.offset         = offsetof(struct qmi_wlanfw_host_cap_req_msg_v01,
+					   mlo_num_chips_valid),
+	},
+	{
+		.data_type      = QMI_UNSIGNED_1_BYTE,
+		.elem_len       = 1,
+		.elem_size      = sizeof(u8),
+		.array_type     = NO_ARRAY,
+		.tlv_type       = 0x25,
+		.offset         = offsetof(struct qmi_wlanfw_host_cap_req_msg_v01,
+					   mlo_num_chips),
+	},
+	{
+		.data_type      = QMI_OPT_FLAG,
+		.elem_len       = 1,
+		.elem_size      = sizeof(u8),
+		.array_type     = NO_ARRAY,
+		.tlv_type       = 0x26,
+		.offset         = offsetof(struct qmi_wlanfw_host_cap_req_msg_v01,
+					   mlo_chip_info_valid),
+	},
+	{
+		.data_type      = QMI_STRUCT,
+		.elem_len       = QMI_WLANFW_MAX_NUM_MLO_CHIPS_V01,
+		.elem_size      = sizeof(struct qmi_wlanfw_host_mlo_chip_info_s_v01),
+		.array_type     = STATIC_ARRAY,
+		.tlv_type       = 0x26,
+		.offset         = offsetof(struct qmi_wlanfw_host_cap_req_msg_v01,
+					   mlo_chip_info),
+		.ei_array      = qmi_wlanfw_host_mlo_chip_info_s_v01_ei,
+	},
+	{
 		.data_type	= QMI_EOTI,
 		.array_type	= NO_ARRAY,
 		.tlv_type	= QMI_COMMON_TLV_TYPE,
@@ -2550,6 +2667,59 @@ out:
 	return ret;
 }
 
+static int ath12k_qmi_fill_mlo_host_caps(struct ath12k_base *ab,
+					 struct qmi_wlanfw_host_cap_req_msg_v01 *req)
+{
+	struct qmi_wlanfw_host_mlo_chip_info_s_v01 *info;
+	struct ath12k_hw_group *ag = ab->ag;
+	int i, j, link_id;
+
+	mutex_lock(&ag->mutex_lock);
+
+	req->mlo_capable = true;
+	req->mlo_capable_valid = true;
+
+	req->mlo_chip_id = 0;
+	req->mlo_chip_id_valid = true;
+
+	req->mlo_group_id = ag->id;
+	req->mlo_group_id_valid = true;
+
+	req->max_mlo_peer = ATH12K_MAX_MLO_PEER;
+	req->max_mlo_peer_valid = true;
+
+	req->mlo_num_chips = ag->num_chip;
+	req->mlo_num_chips_valid = true;
+
+	link_id = 0;
+	for (i = 0; i < ag->num_chip; i++) {
+		info = &req->mlo_chip_info[i];
+
+		info->chip_id = 0;
+		/* TODO: Need to get the phy count */
+		info->num_local_links = 1;
+
+		ath12k_dbg(ab, ATH12K_DBG_QMI, "MLO num_link %d\n", info->num_local_links);
+
+		for (j = 0; j < info->num_local_links; j++) {
+			info->hw_link_id[j] = link_id;
+			info->valid_mlo_link_id[j] = true;
+
+			ath12k_dbg(ab, ATH12K_DBG_QMI, "MLO link id %d\n",
+				    info->hw_link_id[j]);
+
+			link_id++;
+		}
+	}
+
+	req->mlo_chip_info_valid = true;
+
+	ath12k_dbg(ab, ATH12K_DBG_QMI, "MLO host cap chosen\n");
+
+	mutex_unlock(&ag->mutex_lock);
+	return 0;
+}
+
 static int ath12k_qmi_host_cap_send(struct ath12k_base *ab)
 {
 	struct qmi_wlanfw_host_cap_req_msg_v01 req;
@@ -2589,6 +2759,12 @@ static int ath12k_qmi_host_cap_send(stru
 		req.nm_modem |= SLEEP_CLOCK_SELECT_INTERNAL_BIT;
 	}
 
+	ret = ath12k_qmi_fill_mlo_host_caps(ab, &req);
+	if (ret < 0) {
+		ath12k_warn(ab, "Failed to get MLO capability,err = %d\n", ret);
+		goto out;
+	}
+
 	ret = qmi_txn_init(&ab->qmi.handle, &txn,
 			   qmi_wlanfw_host_cap_resp_msg_v01_ei, &resp);
 	if (ret < 0)
Index: backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/qmi.h
===================================================================
--- backports-20220404-5.4.164-f40abb4788.orig/drivers/net/wireless/ath/ath12k/qmi.h	2022-12-21 11:34:10.775816102 +0530
+++ backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/qmi.h	2022-12-21 11:34:10.771816097 +0530
@@ -219,7 +219,7 @@ struct qmi_wlanfw_qdss_trace_mode_resp_m
 	struct qmi_response_type_v01 resp;
 };
 
-#define QMI_WLANFW_HOST_CAP_REQ_MSG_V01_MAX_LEN		261
+#define QMI_WLANFW_HOST_CAP_REQ_MSG_V01_MAX_LEN		351
 
 struct qmi_wlanfw_m3_dump_upload_done_req_msg_v01 {
 	u32 pdev_id;
@@ -243,6 +243,15 @@ struct qmi_wlanfw_m3_dump_upload_done_re
 #define QMI_WLANFW_MAX_PLATFORM_NAME_LEN_V01		64
 #define QMI_WLANFW_MAX_HOST_DDR_RANGE_SIZE_V01		3
 #define QDSS_ETR_MEM_REGION_TYPE                        0x6
+#define QMI_WLANFW_MAX_NUM_MLO_LINKS_PER_CHIP_V01	2
+#define QMI_WLANFW_MAX_NUM_MLO_CHIPS_V01		3
+
+struct qmi_wlanfw_host_mlo_chip_info_s_v01 {
+	u8 chip_id;
+	u8 num_local_links;
+	u8 hw_link_id[QMI_WLANFW_MAX_NUM_MLO_LINKS_PER_CHIP_V01];
+	u8 valid_mlo_link_id[QMI_WLANFW_MAX_NUM_MLO_LINKS_PER_CHIP_V01];
+};
 
 struct qmi_wlanfw_host_ddr_range {
 	u64 start;
@@ -297,6 +306,15 @@ struct qmi_wlanfw_host_cap_req_msg_v01 {
 	u8 mlo_capable;
 	u8 mlo_chip_id_valid;
 	u16 mlo_chip_id;
+	u8 mlo_group_id_valid;
+	u8 mlo_group_id;
+	u8 max_mlo_peer_valid;
+	u16 max_mlo_peer;
+	u8 mlo_num_chips_valid;
+	u8 mlo_num_chips;
+	u8 mlo_chip_info_valid;
+	struct qmi_wlanfw_host_mlo_chip_info_s_v01
+		mlo_chip_info[QMI_WLANFW_MAX_NUM_MLO_CHIPS_V01];
 };
 
 struct qmi_wlanfw_host_cap_resp_msg_v01 {
